client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,306 @@
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.SocketException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.TrustManagerFactory;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.pravega.shared.protocol.netty.AppendBatchSizeTracker.MAX_BATCH_TIME_MILLIS;
+
+@Slf4j
+public class TcpClientConnection implements ClientConnection {
+
+    static final int TCP_BUFFER_SIZE = 256 * 1024;
+    
+    private final Socket socket;
+    private final CommandEncoder encoder;
+    private final ConnectionReader reader;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final PravegaNodeUri location;
+    private final Runnable onClose;
+    private final ScheduledFuture<?> timeoutFuture;
+   
+    private TcpClientConnection(Socket socket, CommandEncoder encoder, ConnectionReader reader, PravegaNodeUri location,
+                                Runnable onClose, ScheduledExecutorService executor) {
+        this.socket = checkNotNull(socket);
+        this.encoder = checkNotNull(encoder);
+        this.reader = checkNotNull(reader);
+        this.location = checkNotNull(location);
+        this.onClose = onClose;
+        this.timeoutFuture = executor.scheduleWithFixedDelay(new TimeoutBatch(encoder),
+                                                             MAX_BATCH_TIME_MILLIS,
+                                                             MAX_BATCH_TIME_MILLIS,
+                                                             TimeUnit.MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    static class ConnectionReader {
+        
+        private final String name;
+        private final InputStream in;
+        private final ReplyProcessor callback;
+        private final ScheduledExecutorService thread;
+        private final AppendBatchSizeTracker batchSizeTracker;
+        private final AtomicBoolean stop = new AtomicBoolean(false);
+
+        public ConnectionReader(String name, InputStream in, ReplyProcessor callback,
+                                AppendBatchSizeTracker batchSizeTracker) {
+            this.name = name;
+            this.in = in;
+            this.callback = callback;
+            this.thread = ExecutorServiceHelpers.newScheduledThreadPool(1, "Reading from " + name);
```

Comment:
We should find a better thread pool name.

Reply:
And name it "executor" or "threadPool" since it's not really a thread. Or If you really want a single thread, then just create a thread, not a pool.

Reply:
Done